crates/edr_provider/src/requests/validation.rs

```diff
@@ -126,3 +127,102 @@ pub fn validate_transaction_spec(
 
     Ok(())
 }
+
+pub fn validate_call_request(
+    spec_id: SpecId,
+    call_request: &CallRequest,
+    block_spec: &Option<BlockSpec>,
+) -> Result<(), ProviderError> {
+    if let Some(ref block_spec) = block_spec {
+        validate_post_merge_block_tags(spec_id, block_spec)?;
+    }
+
+    validate_transaction_and_call_request(
+        spec_id,
+        <&CallRequest as Into<SpecValidationData<'_>>>::into(call_request),
+    )
+}
+
+pub fn validate_transaction_and_call_request<'a>(
+    spec_id: SpecId,
+    validation_data: impl Into<SpecValidationData<'a>>,
+) -> Result<(), ProviderError> {
+    validate_transaction_spec(spec_id, validation_data.into()).map_err(|err| match err {
+        ProviderError::UnsupportedAccessListParameter {
+            minimum_hardfork, ..
+        } => ProviderError::InvalidArgument(format!("\
+Access list received but is not supported by the current hardfork. 
+
+You can use them by running Hardhat Network with 'hardfork' {minimum_hardfork:?} or later.
+        ")),
+        ProviderError::UnsupportedEIP1559Parameters {
+            minimum_hardfork, ..
+        } => ProviderError::InvalidArgument(format!("\
+EIP-1559 style fee params (maxFeePerGas or maxPriorityFeePerGas) received but they are not supported by the current hardfork.
+
+You can use them by running Hardhat Network with 'hardfork' {minimum_hardfork:?} or later.
+        ")),
+        err => err,
+    })
+}
+
+pub fn validate_eip3860_max_initcode_size(
+    spec_id: SpecId,
+    allow_unlimited_contract_code_size: bool,
+    to: &Option<Address>,
+    data: &Bytes,
+) -> Result<(), ProviderError> {
+    if spec_id < SpecId::SHANGHAI || to.is_some() || allow_unlimited_contract_code_size {
+        return Ok(());
+    }
+
+    if data.len() > edr_evm::MAX_INITCODE_SIZE {
+        return Err(ProviderError::InvalidArgument(format!("
+Trying to send a deployment transaction whose init code length is {}. The max length allowed by EIP-3860 is {}.
+
+Enable the 'allowUnlimitedContractSize' option to allow init codes of any length.", data.len(), edr_evm::MAX_INITCODE_SIZE)));
+    }
+
+    Ok(())
+}
+
+pub enum ValidationBlockSpec<'a> {
+    PreEip1898(&'a PreEip1898BlockSpec),
+    PostEip1898(&'a BlockSpec),
+}
+
+impl<'a> From<&'a PreEip1898BlockSpec> for ValidationBlockSpec<'a> {
+    fn from(value: &'a PreEip1898BlockSpec) -> Self {
+        Self::PreEip1898(value)
+    }
+}
+
+impl<'a> From<&'a BlockSpec> for ValidationBlockSpec<'a> {
+    fn from(value: &'a BlockSpec) -> Self {
+        Self::PostEip1898(value)
+    }
+}
+
+pub fn validate_post_merge_block_tags<'a>(
```

Comment:
I added this check for all methods taking a block spec as argument, not just the ones checked by the integration tests.